pkg/controller/meta_controller.go

```diff
@@ -0,0 +1,145 @@
+// Copyright 2023 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controller
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	pb "github.com/risingwavelabs/risingwave-operator/pkg/controller/proto"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/config"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+)
+
+// MetaPodController reconciles a Pod object.
+type MetaPodController struct {
+	client.Client
+	// Scheme *runtime.Scheme
+}
+
+// +kubebuilder:rbac:groups=core,resources=pods,verbs=get;list;watch;update;patch
+
+const (
+	metaLeaderLabel = "meta-is-leader"
+)
+
+// metaIsLeader sends a heartbeat to a meta node at ip:port. If meta responds it is the leader.
+func (r *MetaPodController) metaIsLeader(ctx context.Context, ip string, port uint) bool {
+	addr := fmt.Sprintf("%s:%v", ip, port)
+	log := log.FromContext(ctx)
+
+	// TODO: Do we need to make this connection secure?
+	conn, err := grpc.Dial(addr, grpc.WithTransportCredentials(insecure.NewCredentials()))
```

Comment:
Pass the `ctx` down to ensure the cancellation works. Also, please give it a timeout to avoid getting stuck on hanging connections.

Reply:
Are you sure that I need to pass it down? This looks just like in the documentation https://pkg.go.dev/context#WithTimeout 

```
func slowOperationWithTimeout(ctx context.Context) (Result, error) {
	ctx, cancel := context.WithTimeout(ctx, 100*time.Millisecond)
	defer cancel()  // releases resources if slowOperation completes before timeout elapses
	return slowOperation(ctx)
}
```

Reply:
I am passing the `context` down to the function now. I am not sure that I need to give a timeout to the individual `grpc` request. Should the [context.WithTimeout](https://pkg.go.dev/context#WithTimeout) not provide the timeout for us?